sklearn/impute/_base.py

```diff
@@ -519,9 +519,16 @@ def transform(self, X):
             if invalid_mask.any():
                 missing = np.arange(X.shape[1])[invalid_mask]
                 if self.verbose != "deprecated" and self.verbose:
-                    warnings.warn(
-                        "Skipping features without observed values: %s" % missing
-                    )
+                    # use feature names warning if features are provided
+                    if hasattr(self, "feature_names_in_"):
+                        warnings.warn(
+                            "Skipping features without observed values: %s"
+                            % self.feature_names_in_[missing]
+                        )
+                    else:
+                        warnings.warn(
+                            "Skipping features without observed values: %s" % missing
+                        )
```

Comment:
I think this could be improved further as:


```python
            if invalid_mask.any():
                invalid_features = np.arange(X.shape[1])[invalid_mask]
                if self.verbose != "deprecated" and self.verbose:
                    # use feature names warning if features are provided
                    if hasattr(self, "feature_names_in_"):
                        invalid_features = self.feature_names_in_[invalid_features]
                    warnings.warn(
                        f"Skipping features without any observed values: {invalid_features}. "
                        "At least one non-missing value is needed for imputation with "
                        f"strategy={self.strategy}."
                    )
```

Beware that this code change is untested and probably requires black reformatting.

Reply:
Yes, thank you for your improvement. I will incorporate it in the next commit.